[PAS-109] Update Profile Dropdown Sign in/Sign out behaviour

BraidenCutforth opened the pull request.

BraidenCutforth committed:
wip

BraidenCutforth committed:
Remove unused vars, improve sign out behaviour

BraidenCutforth committed:
Update useComponentVisible hook

BraidenCutforth committed:
Change the structure of the Generic Dropdown (Inheritance -> Composition)

daelynj commented on src/components/GenericDropdown.tsx:
```diff
@@ -15,38 +14,87 @@ const DropdownButton = styled.button`
+		if (buttonRef.current != null) {
+			const rect = buttonRef.current.getBoundingClientRect();
+			setTop(rect.bottom);
+		}
```
This would be a nice place to extract a function or even encapsulate the entire `useEffect` in a custom hook with a good name, I'm not too sure what the purpose of this code is by reading it at first glance. Remember though that if we decide to pull it out (not as a custom hook) that whatever function is defined will need a `useCallback` if it's defined in the component scope so it's not redefined every re-render. If it's defined outside the component scope it does not need a `useCallback`. Another option is defining a function inside `useEffect` and calling it.

Edit: reading further, if this was a custom hook we would also be able to encapsulate the `[top, setTop]` state as well, further cleaning up this code.

BraidenCutforth replied:
So I agree about cleaning this up, and I think it was less confusing when there were more components as the context of the code provided a lot of insight. 

I think a custom hook is interesting, but I don't see it being particularly generic. Meaning no other component needs to be displayed floating below another component. That being said, what are your thoughts about defining a hook outside(above) the component? 

I think what I have written now is more clear, but probably some room for improvement. I think a custom hook though is kind of overkill for this situation...

daelynj replied:
It's good to think in terms of reuse; it would be fine to pull the state and lifecycle logic into its own hook for the sake of encapsulation and as a separation of concerns. `GenericDropdown` shouldn't necessarily be concerned with the code that makes the dropdown work, ie. the hooks.

On one hand we have view logic, the code that handles displaying data with styling. On the other hand we have a form of business logic, the code that handles state and manipulation of the viewport using the `Element` class. These two things work hand-in-hand but they don't necessarily need to be defined in the same place.

If testing was a concern, separating the hook would also allow us to test the hook in isolation. As it stands, tests for `GenericDropdown` would have to implicitly test the lifecycle/state logic.

What I personally wouldn't do is define the hook in the same file; this is a bit of an anti-pattern, much like having two components built in one file.

The extra level of abstraction could be seen as overkill sure, but at first glance I personally don't know what this lifecycle hook is doing. At the very least it wouldn't hurt anyone and has the added benefit of possibly cleaning up the code and improving readability.

daelynj commented on src/hooks/useComponentVisible.tsx:
```diff
@@ -4,22 +4,25 @@ export default function useComponentVisible(
-
 	useEffect(() => {
+		const handleClickOutside = (event: MouseEvent) => {
+			if (ref.current && !ref.current.contains(event.target as Node)) {
```
`if (ref.current && !ref.current.contains(event.target as Node))`

Maybe we can extract this into a variable with a good name and place it in the `if` statement so the purpose of the conditional is clear.

If it's short enough it might even read nicer as a one liner:
```js
if (theConditional) setComponentVisible(false);
```

daelynj commented on src/hooks/useSignedIn.tsx:
```diff
@@ -0,0 +1,26 @@
+	const checkAuthenticated = () => {
+		Auth.currentAuthenticatedUser()
+			.then((user) => {
+				console.log(user);
```
I assume this was used for debugging, we don't want it in master.

Ideally ESLint (or other linter) should be catching this in a pre-commit git hook as well as part of a linting CI check.

here's the rule:
https://eslint.org/docs/rules/no-console

BraidenCutforth replied:
We had console.logs as placeholders. When we are ready for production, I will enforce this. Ticket made!

daelynj reviewed (changes requested):
Great changes. I think we'll at least want to remove the `console.log` before we merge. If you request my review again and I'm not getting to it just DM me in case I didn't see the re-request.

daelynj commented:
Try to keep commit titles at or below 50 characters so we don't end up with the title wrapping into the commit message like this. Anything we can do to make reading our commits easier for reviewers is a good idea. The best way to avoid this is to never use the `-m` flag while committing as there is no warning when your title is too long.


![Screen Shot 2021-04-19 at 2 38 12 PM](https://user-images.githubusercontent.com/28407708/115306583-e51d7000-a11c-11eb-97d4-ac7ee2032ba5.png)

--------------

Best practices call for us to write a descriptive title of the change and then explain why we made it.
```
Render List Items as children of GenericDropdown

By following the best practices laid out in the React documentation for
Composition vs. Inheritance we can decrease the depth of our component
rendering tree while providing a higher level of inversion of control to
consumers of the GenericDropdown component.

References:
https://reactjs.org/docs/composition-vs-inheritance.html
```

BraidenCutforth committed:
PAS-109: Clean up Generic Dropdown

- Extract function for setting dropdown list position
- Extract variable for detecting outside click (useComponentVisible
hook)
- Remove console log from useSignedIn hook

daelynj commented on src/components/GenericDropdown.tsx:
```diff
@@ -15,38 +14,90 @@ const DropdownButton = styled.button`
+
+	useEffect(() => {
+		const getMenuButtonBottom = () => {
+			if (buttonRef.current != null) {
```
What does `buttonRef.current != null` mean? Could we pull this into a variable?

BraidenCutforth replied:
I mean this seems like overkill haha, but I extracted `buttonRef.current` (DOM element) as `menuButton` .

```typescript
const getMenuButtonBottom = () => {
	const menuButton = buttonRef.current;
	if (menuButton != null) {
		return menuButton.getBoundingClientRect().bottom;
	}
	return HIDE_LIST;
};
```
And at this point I think it is very clear, without giving the return type a name.

daelynj replied:
I don't think `buttonRef.current` is what's unclear, rather `buttonRef.current != null`, is this checking the existence of the button? or if the button is enabled vs. disabled?

I'd rather see something like
```js
const buttonExists = buttonRef.current != null;

if (buttonExists) {
```

BraidenCutforth replied:
I disagree on that one, because I think `menuButton != null` reads very clearly as menuButton is defined.

daelynj commented on src/components/GenericDropdown.tsx:
```diff
@@ -15,38 +14,90 @@ const DropdownButton = styled.button`
+	useEffect(() => {
+		const getMenuButtonBottom = () => {
+			if (buttonRef.current != null) {
+				return buttonRef.current.getBoundingClientRect().bottom;
```
This return value also needs a name, what's its purpose? Would `bottomOfViewport` work?

```js
const getMenuButtonBottom =() => {
  const goodName = buttonRef.current != null;

  if (goodName) {
    const bottomOfViewport = buttonRef.current.getBoundingClientRect().bottom;
    return bottomOfViewport;
  }
}

BraidenCutforth replied:
You know what the return value is based on the name of the function

daelynj commented on src/components/GenericDropdown.tsx:
```diff
@@ -15,38 +14,90 @@ const DropdownButton = styled.button`
+			if (buttonRef.current != null) {
+				return buttonRef.current.getBoundingClientRect().bottom;
+			}
+			return -9999; // Default hide the list
```
A comment here is fine, generally I find it helpful to declare these types of static numbers that have some sort of meaning into a constant, likely at the top of the file like
```js
const HIDE_THE_LIST = -9999;

return HIDE_THE_LIST;
```
This removes the need for a comment.

BraidenCutforth committed:
PAS-109: Refine setting dropdown list location

Samantha-norrie reviewed (approved):
Woo! ✔️

BraidenCutforth committed:
PAS-109: Change DropdownButton to MenuButton

- Rename DropdownButton to MenuButton for clarity

BraidenCutforth committed:
PAS-109: Remove unused import
